lib/Common/Memory/Recycler.inl

```diff
@@ -517,6 +521,13 @@ Recycler::AddMark(void * candidate, size_t byteCount) throw()
     return markContext.AddMarkedObject(candidate, byteCount);
 }
 
+inline bool
+Recycler::AddPreciselyTracedMark(IRecyclerVisitedObject * candidate) throw()
+{
+    // This is never called during parallel marking
+    Assert(this->collectionState != CollectionStateParallelMark);
```

Comment:
Is this ever called during mark or only during rescan? Can we turn this assert into the positive case instead of the negative case?

Reply:
Only during rescan.  I can change the assert.

Reply:
Actually, when changing the assert and running GCStress I find that while called under BackgroundRescan the Recycler's collectionState is Collection_ConcurrentMark... which doesn't have the Collection_Rescan flag.  I think the important aspect of the assert, however, is captured in that a Recycler API which pushes something into a MarkingContext lacks the information necessary to know which MarkingContext in the case of parallel marking.

I made a minor change to the assert to make this clearer:

    // This API cannot be used for parallel marking as we don't have enough information to determine which MarkingContext to use.
    Assert((this->collectionState & Collection_Parallel) == 0);